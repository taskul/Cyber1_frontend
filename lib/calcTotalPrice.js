export default function calcTotalPrice(cart) {
  // tally is our accumulator of reduce and CartItem,
  return cart.reduce((tally, cartItem) => {
    // products could be deleted, but they could still be in your cart
    if (!cartItem.product) return tally;
    return tally + cartItem.quantity * cartItem.product.price;
  }, 0);
}
