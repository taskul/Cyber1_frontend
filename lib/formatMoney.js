export default function formatMoney(amount = 0) {
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2,
  };

  // check if its a clean dollar amount
  // so then $23.00 will be displayed as $23
  if (amount % 100 === 0) {
    options.minimumFractionDigits = 0;
  }

  // Intl.NumberFormat is a built in Node.js function for converting currency
  const formatter = new Intl.NumberFormat('en-US', options);
  // we need to also divide it by 100
  return formatter.format(amount / 100);
}
