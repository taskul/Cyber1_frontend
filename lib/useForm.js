import { useEffect, useState } from 'react';

// we'll take in some initial state adn we'll just default that to an empty object, just in case someone doesn't pass in any default state.
export default function useForm(initial = {}) {
  // create a state object for our inputs
  const [inputs, setInputs] = useState(initial);

  // we set this up to avoid infinite loop in useEffect
  const initialValues = Object.values(initial).join('');

  useEffect(() => {
    // this function runs when the things we are watching change.
    setInputs(initial);
  }, [initialValues]);

  // This will handle the change for our inputs
  // we are setting the entire state to an object, so that way we can work with multiple inputs
  function handleChange(e) {
    // input fields always return strings for values, which will be a problem when we send a string to a GraphQL field that required a number. So here is what we can do:
    // first destructure e.target object and get "type" for type of input field
    let { value, name, type } = e.target;
    // then if type is number, we'll convert the string to a number
    if (type === 'number') {
      value = parseInt(value);
    }

    // if we are dealing with uploading files
    // e.target.files is an array, and what we want is the first value from that array e.target.files[0], and can destructure that array by passing the first item as value in [value] = e.target.files;
    if (type === 'file') {
      [value] = e.target.files;
    }
    setInputs({
      // copy the existing state,
      ...inputs,
      // this way we can dynamically change name of the input field
      [name]: value,
    });
  }
  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    // here we are clearing all the values but keeping the keys
    // Object.fromEntries helps us to turn the array back into an object
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key, value]) => [key, ''])
    );
    setInputs(blankState);
  }

  return { inputs, handleChange, resetForm, clearForm };
}
