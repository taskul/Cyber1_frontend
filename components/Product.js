import Link from 'next/link';
import ItemStyles from './styles/ItemStyles';
import Title from './styles/Title';
import PriceTag from './styles/PriceTag';
import formatMoney from '../lib/formatMoney';
import DeleteProduct from './DeleteProducts';
import AddToCart from './AddToCart';

export default function Product({ product }) {
  return (
    <ItemStyles>
      {/* Javascript optional chaining parameter ?
      product?.photo?.image? question mark at the end checks if that item exists, does product exist, does photo exist, does image exist */}
      <img
        src={product?.photo?.image?.publicUrlTransformed}
        alt={product.name}
      />
      <Title>
        <Link href={`/product/${product.id}`}>{product.name}</Link>
      </Title>
      <PriceTag>{formatMoney(product.price)}</PriceTag>
      <p>{product.description}</p>

      {/* This is another way to do your URLs in Next.Js, and the way we can access the ID is via the query on the page. */}
      <div className="buttonList">
        <Link
          href={{
            pathname: 'update',
            query: {
              id: product.id,
            },
          }}
        >
          Edit &#128393;
        </Link>
        <AddToCart id={product.id} />
        <DeleteProduct id={product.id}>Delete &#128465;</DeleteProduct>
      </div>
    </ItemStyles>
  );
}
