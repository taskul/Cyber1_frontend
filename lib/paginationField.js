import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // tells Apollo that we'll take care of everything

    // any items in the cache are the existing items
    // the second argument is an object that we can destructure to get args and cache
    // --- args are the $skip and $first values we used to pick how many items to skip and display
    // --- cache is the Apollo cache
    read(existing = [], { args, cache }) {
      const { skip, first } = args;
      // read the number of items on the page from the cache
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      // this is the total number of products we have
      const count = data?._allProductsMeta?.count;
      // calculte the page we are currently on
      const page = skip / first + 1;
      // how many total pages there are
      const pages = Math.ceil(count / first);
      //   console.log('EXISTING', existing);

      // check if we have existing items on the page
      // we pop on filter at the end, so that way if someting is undefined, only items that are not undefined will be returned
      const items = existing.slice(skip, skip + first).filter((x) => x);
      // if there are items and there aren't enough items to satisfy how may awere requested
      // and we are on the last page, THEN just send it
      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      if (items.length !== first) {
        // we don't have any items, we must go to the network to fetch them
        return false;
      }
      // if there are items, just return them from the cache, and we don't need to go to the network
      if (items.length) {
        // if there are items in the cache we are going to send them to the Apollo
        return items;
      }

      return false; // fallback to network
      // When Apollo queries the products first thing it does is asks the read function for those items
      // Here we can do one of two things:
      // First Option: return the items because they are already in the cache
      // Other Option: return false from here, which will do network request, if there is nothing in cache then make a network request
    },
    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      // this runs when Apollo client comes back from the network with our products
      //   console.log(`mergin items from the network${incoming.length}`);
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      //   console.log(merged);
      // finally we return the merged items form the cache
      // once the merge function returns items, the read function runs again
      return merged;
    },
  };
}
