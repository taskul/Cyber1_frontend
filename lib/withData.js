import { ApolloClient, ApolloLink, InMemoryCache } from '@apollo/client';
import { onError } from '@apollo/link-error';
import { getDataFromTree } from '@apollo/client/react/ssr';
import { createUploadLink } from 'apollo-upload-client';
import withApollo from 'next-with-apollo';
import { endpoint, prodEndpoint } from '../config';
import paginationField from './paginationField';

function createClient({ headers, initialState }) {
  // we create a new Apollo Client
  return new ApolloClient({
    // then we insert links
    // first link is an error handling link. It takes two different types of errors that could possibly happen in your GraphQL requests - graphQLErrors, networkError.
    // some examples of error are your password is wrong or you requested a field that does not exist.
    // or you may request an image on a product and it will say did you mean images? or did you mean photos?
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors)
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        // the second is a network error, that would activate when the backend is down, or there is a CORS issue.
        if (networkError)
          console.log(
            `[Network error]: ${networkError}. Backend is unreachable. Is it running?`
          );
      }),
      // this uses apollo-link-http under the hood, so all the options here come from that package
      // this is responsible for fetching data as well as making post requests, esentially all CRUD events.
      // Inside of React, Apollo is going to give up both query and mutation hooks that will allow us to do our CRUD requests.
      createUploadLink({
        // where is the GraphQL endpoin
        uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
        // here we indicate if we should send cookies with each request, and in our case, yes we do want to send cookies in order to make sure the user is logged in, and if they are check what they are actually allowed to see (we have user roles),
        fetchOptions: {
          credentials: 'include',
        },
        // pass the headers along from this request. This enables Server Side Rendering (SSR) with logged in state
        // this allows us to render on the server side all of the logged in states.
        // we are also going to store JWT in cookies so then it can be sent to the server. Storing JWT in localStorage does not allow server to access it
        headers,
      }),
    ]),
    // where will we store the cache, and here we set it up to store it in memory.
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            // TODO: We will add this together!
            // the way we get data from the queries it is already divided into pages using $skip and $first and then it is added to cache. However deleting items is hard because items don't shift on pages because of how they are caches. So we need to control how things are cached here
            allProducts: paginationField(),
          },
        },
      },
      // restore - what that does is that is because we're initially rendering these views on the server, we need to hot potato all the data that was collected on the server Apollo Clinet, and we need to give it to the hydration of the client. And this is what it's doing, it's saying if there is any initial state, then restore it, otherwise, go ahead.
    }).restore(initialState || {}),
  });
}

// getDataFromTree
// withApollo allows us to crawl all of our pages and components, and look for any queries that we have. So queries for products, users, and orders. It will wait for all that data to be fetched before the server send the HTML from the server to the Client.
export default withApollo(createClient, { getDataFromTree });
