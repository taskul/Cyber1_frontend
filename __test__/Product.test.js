import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import Product from '../components/Product';
import { fakeItem } from '../lib/testUtils';

const product = fakeItem();

describe('<Product/>', () => {
  it('renders out the price tag and title', () => {
    const { container, debug } = render(
      // we use MockedProvider to fake having Apollo in our tests
      <MockedProvider>
        <Product product={product} />
      </MockedProvider>
    );
    // debug shows us the html components we are rendering
    // debug();
    const priceTag = screen.getByText('$50');
    // we can debug specific html components rendered
    // debug(priceTag);
    expect(priceTag).toBeInTheDocument();
    const link = container.querySelector('a');
    // debug(link);
    expect(link).toHaveAttribute('href', '/product/abc123');
    expect(link).toHaveTextContent(product.name);
  });

  // create a snap shot of the Product HTML render, when we change Product component we need to update snapshot. When Jest is running and you get an error of snapshot does not match, press "u" to update snapshot
  it('Renders and matches the snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <Product product={product} />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders the image properly', () => {
    const { container, debug } = render(
      <MockedProvider>
        <Product product={product} />
      </MockedProvider>
    );
    // grab the image, out alt for image is the product.name
    const img = screen.getByAltText(product.name);
    expect(img).toBeInTheDocument();
  });
});
