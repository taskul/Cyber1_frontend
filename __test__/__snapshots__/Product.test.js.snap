// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product/> Renders and matches the snapshot 1`] = `
<div>
  <div
    class="ItemStyles-sc-16pk14u-0 ftBZFq"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <h3
      class="Title-sc-16nq74k-0 kBnoQb"
    >
      <a
        href="/product/abc123"
      >
        dogs are best
      </a>
    </h3>
    <span
      class="PriceTag-sc-nwbk6t-0 zDzVV"
    >
      $50
    </span>
    <p>
      dogs
    </p>
    <div
      class="buttonList"
    >
      <a
        href="/update?id=abc123"
      >
        Edit ðŸ–‰
      </a>
      <button
        style="cursor: pointer;"
        type="button"
      >
        Add
         To Cart ðŸ‘œ
      </button>
      <button
        style="cursor: pointer;"
        type="button"
      >
        Delete ðŸ—‘
      </button>
    </div>
  </div>
</div>
`;
